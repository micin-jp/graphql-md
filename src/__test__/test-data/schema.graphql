type Query {
  getUser(id: ID!): User
  getPost(id: ID!): Post
  getPosts: [Post!]!
}

type Mutation {
  createUser(input: CreateUserInput!): User
  createPost(input: CreatePostInput!): Post
}

type User {
  id: ID!
  name: String!
  email: String!
  posts: [Post!]!
}

type Post {
  id: ID!
  title: String!
  content: String!
  author: User!
}

enum Role {
  ADMIN
  USER
}

union SearchResult = User | Post

interface Node {
  id: ID!
}

input CreateUserInput {
  name: String!
  email: String!
  password: String!
  role: Role = USER
}

input CreatePostInput {
  title: String!
  content: String!
  authorId: ID!
}
