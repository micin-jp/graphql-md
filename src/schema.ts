import { readFileSync } from 'fs'

import {
  buildSchema,
  GraphQLSchema,
  GraphQLEnumType,
  GraphQLUnionType,
  GraphQLObjectType,
  GraphQLInputObjectType,
} from 'graphql'
import { Config } from './config'

const DEFAULT_GRAPHQL_SCALARS = [
  'String',
  'Int',
  'Float',
  'BigInt',
  'Boolean',
  'DateTime',
  'JSON',
  'JSONObject',
]

export type SchemaData = {
  queries?: GraphQLObjectType
  mutations?: GraphQLObjectType
  types: {
    objects: GraphQLObjectType[]
    enums: GraphQLEnumType[]
    unions: GraphQLUnionType[]
    inputs: GraphQLInputObjectType[]
  }
}

export const loadSchema = (config: Config): SchemaData => {
  // Read the schema file
  const schemaString = readFileSync(config.schemaPath, 'utf8')

  // Build the schema object
  const schema: GraphQLSchema = buildSchema(schemaString)

  const queries = schema.getQueryType() ?? undefined
  const mutations = schema.getMutationType() ?? undefined
  const types = Object.values(schema.getTypeMap())

  const objects = types
    .filter((t): t is GraphQLObjectType => t instanceof GraphQLObjectType)
    .filter((t) => !DEFAULT_GRAPHQL_SCALARS.includes(t.name))
    .filter((t) => !t.name.startsWith('__'))
    .filter((t) => t.name !== 'Query' && t.name !== 'Mutation')
  const enums = types.filter(
    (t): t is GraphQLEnumType => t instanceof GraphQLEnumType
  )
  const unions = types.filter(
    (t): t is GraphQLUnionType => t instanceof GraphQLUnionType
  )
  const inputs = types.filter(
    (t): t is GraphQLInputObjectType => t instanceof GraphQLInputObjectType
  )

  return {
    queries,
    mutations,
    types: {
      objects,
      enums,
      unions,
      inputs,
    },
  }
}