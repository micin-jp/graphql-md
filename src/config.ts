import { readFileSync } from 'fs'
import { parse } from 'yaml'

import { z } from 'zod'

export const Config = z.object({
  title: z.string().default('API Documentation'),
  description: z.string().default(''),
  schemaPath: z.string().default('schema.graphql'),
  outputPath: z.string().default('graphql-schema.md'),
})

export type Config = z.infer<typeof Config>

export const loadConfig = (configPath: string): Config => { 
  const yamlStr = readFileSync(configPath, 'utf8')
  const parsed = parse(yamlStr)
  const validationResult = Config.safeParse(parsed)
  if (validationResult.success) {
    return validationResult.data
  }
  throw new Error(validationResult.error.message)
}