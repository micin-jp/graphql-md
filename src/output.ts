import { writeFile } from 'fs'

import { Config } from "./config";

export const output = async (source: string, config: Config): Promise<void> => {
  await writeFile(config.outputPath, source, (err) => {
    if (err) throw err
  })
}