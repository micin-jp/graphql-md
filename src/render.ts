
import {
  GraphQLEnumType,
  GraphQLUnionType,
  GraphQLObjectType,
  GraphQLInputObjectType,
  GraphQLArgument,
} from 'graphql'
import { SchemaData } from './schema'
import { Config } from './config'

const toPascalCase = (str: string) => {
  return str
    .split(/[-_]/)
    .map((s) => s[0].toUpperCase() + s.slice(1))
    .join('')
}

const toSectionLink = (str: string) => {
  return `[${str}](#${toPascalCase(str)})`
}

const wrapString = (str: string) => str.replace(/\n/g, '<br>')

const argsToMd = (args: ReadonlyArray<GraphQLArgument>): string => {
  return args.map((arg) => `${arg.name}: ${toSectionLink(arg.type.toString())}`).join('<br>')
}

const getTypeMarkdown = (
  type:
    | GraphQLObjectType
    | GraphQLEnumType
    | GraphQLUnionType
    | GraphQLInputObjectType
): string => {
  let markdown = `#### ${type.name}\n\n`
  if (type.description) {
    markdown += `${type.description}\n\n`
  }
  if (
    type instanceof GraphQLObjectType ||
    type instanceof GraphQLInputObjectType
  ) {
    markdown += '| Field | Type | Description |\n'
    markdown += '| ----- | ---- | ----------- |\n'
    for (const field of Object.values(type.getFields())) {
      markdown += `| ${field.name} | ${field.type} | ${wrapString(
        field.description ?? ''
      )} |\n`
    }
  } else if (type instanceof GraphQLEnumType) {
    markdown += '| Value | Description |\n'
    markdown += '| ----- | ----------- |\n'
    for (const value of Object.values(type.getValues())) {
      markdown += `| ${value.name} | ${wrapString(value.description ?? '')} |\n`
    }
  } else if (type instanceof GraphQLUnionType) {
    markdown += '| Type | Description |\n'
    markdown += '| ---- | ----------- |\n'
    for (const value of Object.values(type.getTypes())) {
      markdown += `| ${value.name} | ${wrapString(value.description ?? '')} |\n`
    }
  }
  return markdown
}


export const render = (schemaData: SchemaData, config: Config): string => {
  const title = config.title
  const description = config.description
  // Generate the Markdown documentation
  let markdown = ''

  // Add the title
  markdown += `# ${title}\n\n`
  markdown += `${description}\n\n`
  const {
    queries,
    mutations,
    types: { objects, enums, unions, inputs },
  } = schemaData

  if (queries) {
    // Add the Query types
    markdown += '## Query\n\n'
    markdown += '| name | Type | Description |\n'
    markdown += '| ---- | ---- | ----------- |\n'
    for (const field of Object.values(queries.getFields())) {
      markdown += `| ${field.name} | ${
        toSectionLink(field.type.toString())
      } | ${wrapString(field.description ?? '')} |\n`
    }
  }

  if (mutations) {
    // Add the Mutation types
    markdown += '## Mutation\n\n'
    markdown += '| name | input | Type | Description |\n'
    markdown += '| ---- | -- | -- | ----------- |\n'
    for (const field of Object.values(mutations.getFields())) {
      markdown += `| ${field.name} | ${argsToMd(field.args)} | ${
        toSectionLink(field.type.toString())
      } | ${wrapString(field.description ?? '')} |\n`
    }
  }

  markdown += '## Types \n\n'

  if (objects.length > 0) {
    markdown += '### Objects\n\n'
    for (const o of objects) {
      markdown += getTypeMarkdown(o)
    }
  }

  if (enums.length > 0) {
    markdown += '### Enums\n\n'
    for (const e of enums) {
      markdown += getTypeMarkdown(e)
    }
  }

  if (unions.length > 0) {
    markdown += '## Unions\n\n'
    for (const u of unions) {
      markdown += getTypeMarkdown(u)
    }
  }

  if (inputs.length > 0) {
    markdown += '### Inputs\n\n'
    for (const i of inputs) {
      markdown += getTypeMarkdown(i)
    }
  }

  return markdown
}
