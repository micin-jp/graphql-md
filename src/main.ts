import { loadSchema } from "./schema"
import { render } from "./render"
import { loadConfig } from "./config"
import { output } from "./output"
import { loadArgs } from './args';


export const exec = () => { 
  const args = loadArgs()
  const config =  loadConfig(args.configPath)
  const schemaData = loadSchema(config)
  const markdown = render(schemaData, config)
  output(markdown, config)  
}
