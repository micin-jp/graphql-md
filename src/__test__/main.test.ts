import { writeFile } from "fs";
import { loadConfig } from "../config";
import { exec } from "../main";
import { output } from "../output";
import { render } from "../render";
import { loadSchema } from "../schema";

vi.mock('../args', () => {
  return {
    ...vi.importActual('../args'),
    loadArgs: vi.fn().mockReturnValue({
      configPath: 'src/__test__/test-data/graphql-md.yaml'
    }),
  }
});


describe('Test', () => { 
  it('should be true', () => {
    const c = loadConfig('src/__test__/test-data/graphql-md.yaml')
    const s = loadSchema(c)
    const r = render(s, c)
    expect(r).toMatchSnapshot()
  })
});