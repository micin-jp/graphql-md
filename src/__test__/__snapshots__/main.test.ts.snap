// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test > should be true 1`] = `
"# API Documentation

test description

## Query

| name | Type | Description |
| ---- | ---- | ----------- |
| [getUser](#GetUser) | User |  |
| [getPost](#GetPost) | Post |  |
## Mutation

| name | input | Type | Description |
| ---- | -- | -- | ----------- |
| [createUser](#CreateUser) | input: CreateUserInput! | User |  |
| [createPost](#CreatePost) | input: CreatePostInput! | Post |  |
## Types 

### Objects

#### User

| Field | Type | Description |
| ----- | ---- | ----------- |
| id | ID! |  |
| name | String! |  |
| email | String! |  |
| posts | [Post!]! |  |
#### Post

| Field | Type | Description |
| ----- | ---- | ----------- |
| id | ID! |  |
| title | String! |  |
| content | String! |  |
| author | User! |  |
### Enums

#### Role

| Value | Description |
| ----- | ----------- |
| ADMIN |  |
| USER |  |
#### __TypeKind

An enum describing what kind of type a given \`__Type\` is.

| Value | Description |
| ----- | ----------- |
| SCALAR | Indicates this type is a scalar. |
| OBJECT | Indicates this type is an object. \`fields\` and \`interfaces\` are valid fields. |
| INTERFACE | Indicates this type is an interface. \`fields\`, \`interfaces\`, and \`possibleTypes\` are valid fields. |
| UNION | Indicates this type is a union. \`possibleTypes\` is a valid field. |
| ENUM | Indicates this type is an enum. \`enumValues\` is a valid field. |
| INPUT_OBJECT | Indicates this type is an input object. \`inputFields\` is a valid field. |
| LIST | Indicates this type is a list. \`ofType\` is a valid field. |
| NON_NULL | Indicates this type is a non-null. \`ofType\` is a valid field. |
#### __DirectiveLocation

A Directive can be adjacent to many parts of the GraphQL language, a __DirectiveLocation describes one such possible adjacencies.

| Value | Description |
| ----- | ----------- |
| QUERY | Location adjacent to a query operation. |
| MUTATION | Location adjacent to a mutation operation. |
| SUBSCRIPTION | Location adjacent to a subscription operation. |
| FIELD | Location adjacent to a field. |
| FRAGMENT_DEFINITION | Location adjacent to a fragment definition. |
| FRAGMENT_SPREAD | Location adjacent to a fragment spread. |
| INLINE_FRAGMENT | Location adjacent to an inline fragment. |
| VARIABLE_DEFINITION | Location adjacent to a variable definition. |
| SCHEMA | Location adjacent to a schema definition. |
| SCALAR | Location adjacent to a scalar definition. |
| OBJECT | Location adjacent to an object type definition. |
| FIELD_DEFINITION | Location adjacent to a field definition. |
| ARGUMENT_DEFINITION | Location adjacent to an argument definition. |
| INTERFACE | Location adjacent to an interface definition. |
| UNION | Location adjacent to a union definition. |
| ENUM | Location adjacent to an enum definition. |
| ENUM_VALUE | Location adjacent to an enum value definition. |
| INPUT_OBJECT | Location adjacent to an input object type definition. |
| INPUT_FIELD_DEFINITION | Location adjacent to an input object field definition. |
## Unions

#### SearchResult

| Type | Description |
| ---- | ----------- |
| User |  |
| Post |  |
### Inputs

#### CreateUserInput

| Field | Type | Description |
| ----- | ---- | ----------- |
| name | String! |  |
| email | String! |  |
| password | String! |  |
| role | Role |  |
#### CreatePostInput

| Field | Type | Description |
| ----- | ---- | ----------- |
| title | String! |  |
| content | String! |  |
| authorId | ID! |  |
"
`;
